### ISSUE TITLE
Migration table failure against Oracle

### ISSUE THREAD

**Issue by dpassen1** [yogthos/migratus#43]
https://github.com/yogthos/migratus/blob/master/src/migratus/database.clj#L213 uses 'BIGINT' for the Migrations table. This is not a valid datatype in Oracle land.

**Comment by yogthos** [comment]
Yeah, looks like you would either need to detect the database type and use the correct type or allow passing custom types here.

**Comment by yogthos** [comment]
I guess in the meantime, a workaround would be to manually create the table before running migrations.

**Comment by dpassen1** [comment]
I can confirm that this works.

**Comment by alvinfrancis** [comment]
There's a secondary problem here even when manually creating the table, namely there no JDBC mappings to a `java.lang.Long` using the Oracle driver.  Best mapping for the Postgres `BIGINT` in Oracle would be a `NUMBER`, but that's returned by JDBC as a `java.math.BigDecimal`.  This causes `migratus.core/uncompleted-migrations` to fail its check because `migratus.protocols/id` returns a `java.lang.Long`; a `java.math.BigDecimal` does not equal a `java.lang.Long` even with the same value.

Perhaps it would be better to keep ids as some kind of string type column (e.g. `VARCHAR`)? I'd assume JDBC always maps those to `java.lang.String`.  Either that or add coercion in `migratus.database/completed-ids*`to ensure the ids returned are longs.

**Comment by yogthos** [comment]
Using strings for ids should be fine, although that would break backwards compatibility. This looks like an existing issue as well. Perhaps it might be better to just explicitly cast to `long` when comparing?

**Comment by alvinfrancis** [comment]
Yeah.  Considering backwards compatibility, explicit coercion seems to be the easy fix.  It still leaves the bug in Oracle where the migration table cannot be created automatically when it does not exist.  However, since that can be alleviated by manually creating the table, it's less of an issue.

**Comment by yogthos** [comment]
Two options would be to either try to infer the database type from the driver and use the type supported by it, or find a numeric type that most databases support. If the latter is possible, I think that would be ideal.

**Comment by yogthos** [comment]
It looks like ANSI/ISO type should be [double precision](http://www.moreprocess.com/database/sql/data-types-in-sql-defined-by-ansiiso). The [PlatformUtils](http://db.apache.org/ddlutils/api/org/apache/ddlutils/PlatformUtils.html) class from the Apache[DdlUtils](https://db.apache.org/ddlutils/) library could be used to infer the database type otherwise.

**Comment by yogthos** [comment]
Did you try creating the migrations table by hand?

```clojure
(sql/create-table-ddl :schema_migrations
  [[:id "NUMBER" "UNIQUE" "NOT NULL"]
   [:applied "TIMESTAMP" "" ""]
   [:description "VARCHAR(1024)" "" ""]])
```

**Comment by mondaldebanjan** [comment]
This worked :)
hi thanks for the help from the begining, and the quick response yogthos,
